### DISCUSSION TITLE
Accessibility: authentication flows have some major accessibility problems

### DISCUSSION THREAD

**Discussion by ethindp** [keycloak/keycloak#16463] [Archive]
So, I managed to get keycloak working (through Cloudflare, no less), but the authentication page is a mess, accessibility wis. Selecting/duplicating/editing a flow's information/deleting a flow is perfectly fine, but managing sub-flows, steps, and conditions is an utter pain.

The diagram view is pretty much useless. Understandable; it's pretty difficult to make a diagram accessible. But the table view... Isn't actually a table. According to my screen reader, it renders as a list, and this makes it difficult to actually tinker with flows and risks breaking authentication entirely because I have to tab around, and the "press space to drag" stuff gets in the way of me actually determining what flow/step/condition I'm modifying. I need to use something like NVDA's object navigation to figure out the item in question, and that doesn't always work.

As an alternative, I propose replacing the table view with a tree view. This would make navigating among flows significantly easier: sub-flows/conditions/steps could be just tree nodes. You could add accessibility text to identify what the item your on is, and there could be "move up/down" buttons to make dragging them simpler. I know that this probably harkens back to the older way desktop UIs were designed, but this is genuinely simpler to use for screen reader users. (You can make dragging around be a part of the tree view too; I've seen some web apps do this, but I'm unsure how complicated this is, and it most likely requires ARIA.) Furthermore, a tree view would also give the screen reader user an actual picture of what the layout of the flow really is. As it currently stands, this is difficult to determine, and requires quite a bit of guesswork. Such guesswork is dangerous, of course, since if you mess up, you break authentication completely.

The rest of the admin console appears to be pretty well designed, and I have no complaints so far.

**Comment by mrpatrickpilch** [comment]
I agree. In my opinion, the new authentication flow visualization/editing user experience is a **big** step backwards over the previous.
